lib/faktory_worker/connection_manager.ex

```diff
@@ -5,24 +5,60 @@ defmodule FaktoryWorker.ConnectionManager do
 
   alias FaktoryWorker.Connection
 
+  @connection_errors [
+    :closed,
+    :enotconn
+  ]
+
+  defstruct [:opts, :conn]
+
   def start_link(opts \\ []) do
     GenServer.start_link(__MODULE__, opts)
   end
 
-  def send_command(connection, command) do
-    GenServer.call(connection, {:send_command, command})
+  def send_command(connection_manager, command) do
+    GenServer.call(connection_manager, {:send_command, command})
   end
 
   @impl true
   def init(opts) do
-    {:ok, connection} = FaktoryWorker.Connection.open(opts)
-    {:ok, %{conn: connection}}
+    {:ok, %__MODULE__{conn: nil, opts: opts}, {:continue, :setup_connection}}
   end
 
   @impl true
-  def handle_call({:send_command, command}, _, %{conn: connection} = state) do
-    result = Connection.send_command(connection, command)
+  def handle_continue(:setup_connection, %{opts: opts} = state) do
+    {:ok, connection} = Connection.open(opts)
+    {:noreply, %{state | conn: connection}}
+  end
+
+  @impl true
+  def handle_call({:send_command, command}, _, state) do
+    case try_send_command(state, command) do
+      {{:error, reason}, _} when reason in @connection_errors ->
+        error = {:error, "Failed to connect to faktory"}
+        state = %{state | conn: nil}
+
+        {:reply, error, state}
 
-    {:reply, result, state}
+      {result, state} ->
+        {:reply, result, state}
+    end
+  end
+
+  defp try_send_command(%{conn: nil, opts: opts} = state, command) do
+    case Connection.open(opts) do
```

Comment:
Interesting, so this is the path it goes down on first connection attempt rather than using init?

Reply:
It is still using init to start a connection when it boots up using the `handle_continue` so this only gets called once the connection drops off.

I have now refactored this module out into two separate modules so this is now very different and worth another look 🙂